Check deferred nodes after all first passes in an import cycle

gvanrossum opened the pull request:
Might fix #481.

JukkaL commented on test-data/unit/check-modules.test:
```diff
@@ -1175,6 +1175,33 @@ reveal_type(x)
+[file b.py]
+import a
+def g() -> int:
+    return a.y
```
Reveal the type of `a.y` (to verify that it doesn't somehow fall back to `Any`).

JukkaL commented on test-data/unit/check-modules.test:
```diff
@@ -1175,6 +1175,33 @@ reveal_type(x)
+[file a.py]
+import b
+def f() -> int:
+    return b.x
```
Reveal the type of `b.x`.

JukkaL commented on mypy/checker.py:
```diff
@@ -121,55 +121,51 @@ class TypeChecker(NodeVisitor[Type]):
+                                                in options.strict_optional_whitelist)
+
+    def check_first_pass(self) -> None:
+        """Type check the entire file, but defer functions with forward references."""
```
Be more explicit about forward reference mean. In particular, they mean references to definitions whose types haven't been inferred yet.

JukkaL commented:
Looks good to me! This hopefully fixes one of the biggest pain points mypy users with large code bases have had.

I'd like to see some more test cases before merging:

1. Add test case where two passes aren't enough, to verify that we still
generate cannot determine type errors when needed. 

2. Add test case for referring to a decorated function whose type might not be inferred
(another common case where we had trouble before).

3. Add test case where we generate an error only on the second pass (to verify
that the import context is shown correctly). For example, something like this (partial example):
```
def f() -> None:
    a = b.x  # Type of b.x only available during second pass
    a + 1   # Error that we can't report during first pass
```

gvanrossum commented:
I'll add more tests. I also discovered that reveal_type() needs to check for current_node_deferred.

gvanrossum commented:
Rebased, generalized the multi-pass machinery, and added a test that would require 3 passes.

Still needs tests for (2) and (3). Also, there are still some print() calls left.

gvanrossum commented:
I could use a hint for (2).

gvanrossum commented:
I think I have all the required tests now. Jukka, can you double-check?

JukkaL commented on mypy/checker.py:
```diff
@@ -181,21 +187,33 @@ def visit_file(self, file_node: MypyFile, path: str, options: Options) -> None:
+            return False
+        self.errors.set_file(self.path)
+        self.pass_num += 1
+        print('---', self.path, 'pass', self.pass_num + 1, '---')
```
Remove print.

JukkaL commented on mypy/checker.py:
```diff
@@ -181,21 +187,33 @@ def visit_file(self, file_node: MypyFile, path: str, options: Options) -> None:
+        for node, type_name in todo:
+            if node in done:
+                continue
+            print(type_name, '.', node.fullname() or node.name())
```
Remove another print.

JukkaL commented on test-data/unit/check-modules.test:
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
+tmp/a.py:4: error: Cannot determine type of 'x2'
+
+[case testErrorInPassTwo]
+# flags: --hide-error-context
```
Don't hide error context to test the it works correctly for errors reported in the second pass.

JukkaL commented on test-data/unit/check-modules.test:
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
+import a
+x = 1 + 1
+[out]
+tmp/a.py:4: error: Unsupported operand types for + ("int" and "str")
```
Maybe also test this the other way as well (error reported in `b.py` in addition `a.py`.).

JukkaL commented on test-data/unit/check-modules.test:
```diff
@@ -1186,6 +1186,89 @@ reveal_type(x)
+[file a.py]
+import b
+def g() -> None:
+    f()
```
Make sure that the decorator function returns an interesting type (not `Any`) and do `reveal_type(f())` here to verify that decorator got applied correctly.

gvanrossum commented:
Ready for review/merge.

JukkaL commented:
LGTM now -- thanks for the updates!

gvanrossum commented:
OK, rebased and made all requested amendments.
